ArtOfIllusion/src/artofillusion/script/ExecuteScriptWindow.java

```diff
@@ -30,22 +44,104 @@ public class ExecuteScriptWindow extends BFrame
+  // QUESTION should this be static? It's actually the last directory used for scripts, 
+  // is there a reason it is shared among all editing windows? 
+  private static File scriptDir= new File(ArtOfIllusion.TOOL_SCRIPT_DIRECTORY);
+  private String language;
+  private final BButton save;
+  private final BButton executeSelected;
+  private final BButton executeToCursor;
+  private static final int EDITORS_OFFSET = 32;
+  private static ArrayList <String> openedScripts = new ArrayList<String> ();
 
-  public ExecuteScriptWindow(LayoutWindow win)
+    /**
+     * Adds a script path to the recent scripts list. 
+     * This uses a mapping to the current (now) timestamp ; the file paths are in the order of their timestamps, 
+     * so the older is first. The list is truncated to RecentFiles.MAX_RECENT elements. 
+     * @see artofillusion.RecentFiles#MAX_RECENT
+     * @param filePath 
+     */
+    public static void addRecentScript(String filePath)
+    {
+        final Preferences pref = Preferences.userNodeForPackage(ExecuteScriptWindow.class);
+        final String recentFiles[] = pref.get("recentFiles", "").split(File.pathSeparator);
+        final String recentFilesTimes[] = pref.get("recentFilesTimes", "").split(";");
+        Map <Long, String> fileTimes = new TreeMap<> ();
+        for (int fileIndex = 0; fileIndex < recentFiles.length; fileIndex++) {
+            if (!recentFilesTimes [fileIndex].equals ("") // Case of no recent file
+                    && !recentFiles [fileIndex].equals(filePath)) // If the current file already has a timestamp it will be updated below
+                fileTimes.put(Long.valueOf(recentFilesTimes [fileIndex]), recentFiles[fileIndex]);
+        }
+        fileTimes.put (new Date().getTime(), filePath);
+        // truncate the map to the max items selected, removing the first 
+        // (it is the oldest because its timestamp is smaller)
+        if (fileTimes.size() > RecentFiles.MAX_RECENT) {
+            final Long minFileTime = (Long) fileTimes.keySet().toArray() [1];
+            fileTimes = ((TreeMap <Long, String>) fileTimes).tailMap(minFileTime);
+        }
+        pref.put("recentFiles", String.join (File.pathSeparator, fileTimes.values()));
+        final Set<Long> times = fileTimes.keySet();
+        String joinedTimes = "";
+        for (Long time : times) {
+            joinedTimes += ";" + time.toString();
+        }
+        pref.put("recentFilesTimes", joinedTimes.substring(1));
+    }
+    
+    public static String [] getRecentScripts()
+    {
+        final Preferences pref = Preferences.userNodeForPackage(ExecuteScriptWindow.class);
+        // The files are ordered by ascending Date, so we need to revert this to get the most recent in to
+        final java.util.List<String> recentScripts = Arrays.asList (pref.get("recentFiles", "").split(File.pathSeparator));
+        Collections.reverse(recentScripts);
+        return recentScripts.toArray(new String [0]);
+    } 
```

Comment:
... And when combined with this, I think that we can do a bit better. I'd favor a refactor of `RecentFiles` that allows the caller to pass a marker class. I've been working on another project that would also benefit. That way we don't have essentially the same implementation floating around in triplicate.

Reply:
Aaaaaboslutely! I just didn't want to also refactor the Recent Files class (even though I did feel the urge) and keep code changes smaller for a first PR.

Reply:
I actually wonder if a refactored `RecentFiles` might be a smaller change overall? I'm thinking an API something like this:

```
//All these methods take a class argument. Usually `this.getClass` for the calling location
public static void updateRecent(Class interfacePoint, File newestFile)

public static List loadRecent(Class interfacePoint)

public static BMenu createMenu(Class interfacePoint)

@Deprecated(forRemoval=true) // Just for back-compatibility - 
public static void addRecentFile(File file)

@Deprecated(forRemoval=true) // Just for back-compatibility
public static void createMenu(BMenu menu)

Reply:
I like the refactoring idea - and recent files management is a feature that could also be reused for anything in the future - I'm thinkin loading / saving of meshes, procedural textures, renderer settings, whatever comes next. 
One critic, though - I don't like the current code logic of rebuilding the "recent" menu everytime a file is loaded / saved ; I think a "rebuild on-demand" (when the menu is actually opened) would be smarter. 
Maybe leaving it as is for this PR and then grouping a refactored "recent files" class, plus this "rebuild menu on-demand" feature should be a separate PR? Or is it a small enough change that it could be grouped with this PR? What do you think?

Reply:
Not sure which would be better at this point. I do have a project of my own that would benefit from a refactored/extensible RecentFiles implementation. Let's leave it as is, and do the refactor later. (Probably still before next release, but after this PR)